ulib/FStar.List.Tot.Properties.fst

```diff
@@ -285,6 +302,101 @@ let rec map_lemma f l =
     | [] -> ()
     | h::t -> map_lemma f t
 
+(** Properties about unsnoc *)
+
+(** [unsnoc] is the inverse of [snoc] *)
+val lemma_unsnoc_snoc: #a:Type -> l:list a{length l > 0} ->
+  Lemma (requires True)
+    (ensures (let l', x = unsnoc l in snoc l' x == l))
+    [SMTPat (unsnoc l)]
+let lemma_unsnoc_snoc #a l =
+  let l', x = unsnoc l in
+  let l1, l2 = l', [x] in
+  lemma_splitAt_snd_length (length l - 1) l;
+  // assert ((l1, l2) == splitAt (length l - 1) l);
+  let rec aux (l:list a{length l > 0}) :
+    Lemma (let l1, l2 = splitAt (length l - 1) l in
+           append l1 l2 == l) =
+    if length l = 1 then () else aux (tl l) in
+  aux l
+
+(** [unsnoc] followed by [append] can be connected to the same vice-versa. *)
+let rec lemma_unsnoc_append (#a:Type) (l1 l2:list a) :
+  Lemma
+    (requires (length l2 > 0)) // the [length l2 = 0] is trivial
+    (ensures (
+        let as, a = unsnoc (l1 @ l2) in
+        let bs, b = unsnoc l2 in
+        as == l1 @ bs /\ a == b)) =
+  match l1 with
+  | [] -> ()
+  | _ :: l1' -> lemma_unsnoc_append l1' l2
+
+(** [unsnoc] gives you [last] element, which is [index]ed at [length l - 1] *)
+let rec lemma_unsnoc_is_last (#t:Type) (l:list t) :
+  Lemma
+    (requires (length l > 0))
+    (ensures (snd (unsnoc l) == last l /\ snd (unsnoc l) == index l (length l - 1))) =
+  match l with
+  | [_] -> ()
+  | _ -> lemma_unsnoc_is_last (tl l)
+
+(** [index]ing on the left part of an [unsnoc]d list is the same as indexing
+    the original list. *)
+let rec lemma_unsnoc_index (#t:Type) (l:list t) (i:nat) :
+  Lemma
+    (requires (length l > 0 /\ i < length l - 1))
+    (ensures (
+        index (fst (unsnoc l)) i == index l i)) =
+  match i with
+  | 0 -> ()
+  | _ -> lemma_unsnoc_index (tl l) (i - 1)
+
+(** Definition and properties about [split_using] *)
+
+(** [split_using] splits a list at the first instance of finding an
+    element in it. *)
+let rec split_using (#t:Type) (l:list t) (x:t{x `memP` l}) :
+  GTot (r:(list t * list t)) =
+  match l with
+  | [_] -> [], l
+  | a :: as ->
+    if FStar.StrongExcludedMiddle.strong_excluded_middle (a == x) then (
+      [], l
+    ) else (
+      let l1', l2' = split_using as x in
+      a :: l1', l2'
+    )
+
+let rec lemma_split_using (#t:Type) (l:list t) (x:t{x `memP` l}) :
+  Lemma
+    (ensures (
+        let l1, l2 = split_using l x in
+        (length l2 > 0) /\
+        ~(x `memP` l1) /\ (hd l2 == x) /\
+        append l1 l2 == l))
+    [SMTPat (split_using l x)] =
+  match l with
+  | [_] -> ()
+  | a :: as ->
+    if FStar.StrongExcludedMiddle.strong_excluded_middle (a == x)
+    then ()
+    else lemma_split_using (tl l) x
+
+(** Definition of [index_of] *)
+
+(** [index_of l x] gives the index of the leftmost [x] in [l] *)
+let rec index_of (#t:Type) (l:list t) (x:t{x `memP` l}) :
+  GTot (i:nat{i < length l /\ index l i == x}) =
```

Comment:
Same question as above: when do you need this for a non-`eqtype`?

Reply:
Similar answer as earlier, I sometimes need to (in a ghostly sense) get the index of an element which I am holding on to, but it is not an `eqtype` since it is a `pointer`.

Reply:
same as above. Document prominently mentioning the axiom it uses.

Reply:
documented.